File: go/parser.go
package main

import (
	"fmt"
	"os"
	"strings"

	"golang.org/x/exp/slices"
)

type Token struct {
	data string
	line int
	col  int
}

type Parser struct {
	tokens    []Token
	numTokens int
	index     int
}

func createParser(content []byte) *Parser {
	parser := Parser{make([]Token, 0), 0, 0}
	data := string(content[0])
	line := 1
	col := 1
	oldCol := 1
	for i := 1; i < len(content); i += 1 {
		c := content[i]
		if data[len(data)-1] == '\n' || data[len(data)-1] == '\r' || (isSpecialChar(data[len(data)-1]) != isSpecialChar(c)) || c == '#' || c == '[' || c == '(' || c == '\n' || c == '\r' {
			if !strings.Contains(data, "\r") {
				parser.tokens = append(parser.tokens, Token{data, line, oldCol})
				parser.numTokens += 1
			}
			oldCol = col + 1
			if strings.Contains(data, "\n") {
				line += 1
				col = 0
			}
			data = ""
		}
		data += string(c)
		col += 1
	}
	parser.tokens = append(parser.tokens, Token{data, line, col})
	parser.tokens = append(parser.tokens, Token{"\n", line, col})
	return &parser
}

func isSpecialChar(char byte) bool {
	return char == '_' ||
		char == '*' ||
		char == '`' ||
		char == '#' ||
		char == '[' ||
		char == ']' ||
		char == '(' ||
		char == ')' ||
		char == '!'
}

func (parser *Parser) pop() Token {
	parser.index += 1
	return parser.tokens[parser.index-1]
}

func (parser *Parser) peek() Token {
	return parser.tokens[parser.index]
}

func (parser *Parser) accept(token string) *Token {
	if parser.peek().data == token {
		parser.index += 1
		return &parser.tokens[parser.index-1]
	} else {
		return nil
	}
}

func (parser *Parser) expect(token string) {
	if parser.accept(token) == nil {
		top := parser.peek()
		if isSpecialChar(top.data[0]) {
			fmt.Printf("error: %s:%d:%d expected `%s`, got `%s`\n", os.Args[1], top.line, top.col, token, top.data)
		} else {
			fmt.Printf("error: %s:%d:%d expected `%s`, got text\n", os.Args[1], top.line, top.col, token)
		}
		os.Exit(1)
	}
}

func (parser *Parser) parseDocument() []Node {
	retval := make([]Node, 0)
	for parser.index < parser.numTokens-1 {
		if node := parser.parseNode(); node != nil {
			retval = append(retval, node)
		}
	}
	return retval
}

func (parser *Parser) parseNode() Node {
	if parser.accept("#") != nil {
		return parser.parseHeader()
	} else if parser.accept("```") != nil {
		return parser.parseCodeBlock()
	} else if parser.accept("!") != nil {
		return parser.parseImage()
	} else if parser.accept("\n") != nil {
		return nil
	} else {
		return parser.parseParagraph()
	}
}

func (parser *Parser) parseHeader() *Header {
	size := 1
	for parser.accept("#") != nil {
		size += 1
	}
	return &Header{size, parser.parseFormattedText([]string{"\n"})}
}

func (parser *Parser) parseCodeBlock() *CodeBlock {
	text := ""
	for parser.accept("```") == nil {
		text += parser.pop().data
	}
	return &CodeBlock{text}
}

func (parser *Parser) parseImage() *Image {
	parser.expect("[")
	text := parser.pop().data
	parser.expect("]")
	parser.expect("(")
	url := parser.pop().data
	parser.expect(")")
	return &Image{text, url}
}

func (parser *Parser) parseParagraph() *Paragraph {
	return &Paragraph{parser.parseFormattedText([]string{"\n"})}
}

func (parser *Parser) parseFormattedText(bounds []string) []Node {
	children := make([]Node, 0)
	for !slices.Contains(bounds, parser.peek().data) {
		if parser.accept("_") != nil || parser.accept("*") != nil {
			children = append(children, parser.parseItalic(bounds))
		} else if parser.accept("__") != nil || parser.accept("**") != nil {
			children = append(children, parser.parseBold(bounds))
		} else if parser.accept("`") != nil {
			children = append(children, parser.parseCode())
		} else if parser.accept("[") != nil {
			children = append(children, parser.parseLink())
		} else {
			children = append(children, &Text{parser.pop().data})
		}
	}
	return children
}

func (parser *Parser) parseItalic(bounds []string) *Italic {
	newBounds := make([]string, len(bounds))
	copy(newBounds, bounds)
	newBounds = append(newBounds, "*", "_")
	children := parser.parseFormattedText(newBounds)
	if parser.accept("*") == nil {
		parser.expect("_")
	}
	return &Italic{children}
}

func (parser *Parser) parseBold(bounds []string) *Bold {
	newBounds := make([]string, len(bounds))
	copy(newBounds, bounds)
	newBounds = append(newBounds, "**", "__")
	children := parser.parseFormattedText(newBounds)
	if parser.accept("**") == nil {
		parser.expect("__")
	}
	return &Bold{children}
}

func (parser *Parser) parseCode() *Code {
	text := ""
	for parser.accept("`") == nil {
		text += parser.pop().data
	}
	return &Code{text}
}

func (parser *Parser) parseLink() *Link {
	text := parser.pop().data
	parser.expect("]")
	parser.expect("(")
	url := parser.pop().data
	parser.expect(")")
	return &Link{text, url}
}

File: rust/src/parser.rs
use std::collections::HashSet;

use crate::nodes::*;

#[derive(Clone, Debug)]
pub struct Token {
    pub data : String,
    line : i32,
    col  : i32
}

pub struct Parser {
    pub tokens  : Vec<Token>,
    index   : usize
}

pub fn create_parser(contents: String)->Parser {
    let mut parser: Parser = Parser{tokens:Vec::new(), index:0 };
    
    let mut data = String::from("");
    data.push(contents.chars().next().unwrap());
    let mut line = 1;
    let mut col = 1;
    let mut old_col = 1;
    let mut old_c:char = contents.chars().next().unwrap();
    for c in contents.chars().skip(1) {
        if old_c == '\n' || old_c == '\r' || (is_special_char(old_c) != is_special_char(c)) || c == '#' || c == '[' || c == '(' || c == '!' || c == '\n' || c == '\r' {
            parser.tokens.push(Token{data:data, line:line, col:old_col});
            old_col = col + 1;
            if c == '\n' {
                line += 1;
                col = 0;
            }
            data = String::from("");
        }
        if old_c != '#' || !c.is_whitespace() {
            data.push(c);
        }
        col += 1;
        old_c = c;
    }
    parser.tokens.push(Token{data:data, line:line, col:col});
    parser.tokens.push(Token{data:String::from("\n"), line:line, col:col});

    return parser;
}

impl Parser {
    fn pop(&mut self)->Token {
        self.index += 1;
        self.tokens[self.index - 1].clone()
    }

    fn peek(&self)->Token {
        self.tokens[self.index].clone()
    }

    fn accept(&mut self, data:&String)->Option<Token> {
        if self.tokens[self.index].data == *data {
            Some(self.pop())
        } else {
            None
        }
    }

    fn expect(&mut self, data:String) {
        let res = self.accept(&data); 
        match res {
            Some(_) => (),
            None => {
                let top = self.peek();
                if is_special_char(top.data.chars().next().unwrap()) {
                    panic!("error: {}:{} expected `{}` got `{}`", top.line, top.col, data, top.data)
                } else {
                    panic!("error: {}:{} expected `{}` got text", top.line, top.col, data)
                }
            }
        }
    }

    pub fn parse_document(&mut self)->Vec<Box<dyn Node>> {
        let mut retval: Vec<Box<dyn Node>> = Vec::new();
        while self.index < self.tokens.len() - 1 {
            let node = self.parse_node();
            match node {
                Some(n) => retval.push(n),
                None => ()
            }
        }
        retval
    }

    fn parse_node(&mut self)->Option<Box<dyn Node>> {
        if self.accept(&String::from("#")).is_some() {
            Some(Box::new(self.parse_header()))
        } else if self.accept(&String::from("```")).is_some() {
            Some(Box::new(self.parse_code_block()))
        } else if self.accept(&String::from("!")).is_some() {
            Some(Box::new(self.parse_image()))
        } else if self.accept(&String::from("\n")).is_some() {
            None
        } else {
            Some(Box::new(self.parse_paragraph()))
        }
    }

    fn parse_header(&mut self)->Header {
        let mut size = 1;
        while self.accept(&String::from("#")).is_some() {
            size += 1;
        }
        let bounds: HashSet<String> = vec![String::from("\n")].into_iter().collect();
        Header{size:size, children:self.parse_formatted_text(&bounds)}
    }

    fn parse_paragraph(&mut self)->Paragraph {
        let bounds: HashSet<String> = vec![String::from("\n")].into_iter().collect();
        Paragraph{children:self.parse_formatted_text(&bounds)}
    }

    fn parse_code_block(&mut self)->CodeBlock {
        let mut text = String::from("");
        while self.accept(&String::from("```")).is_none() {
            text += &self.pop().data;
        }
        CodeBlock{text}
    }

    fn parse_image(&mut self)->Image {
        self.expect(String::from("["));
        let text = self.pop().data;
        self.expect(String::from("]"));
        self.expect(String::from("("));
        let url = self.pop().data;
        self.expect(String::from(")"));
        Image{text, url}
    }

    fn parse_formatted_text(&mut self, bounds:&HashSet<String>)->Vec<Box<dyn Node>> {
        let mut retval: Vec<Box<dyn Node>> = Vec::new();
        while !bounds.contains(&self.peek().data) {
            if self.accept(&String::from("_")).is_some() || self.accept(&String::from("*")).is_some() {
                retval.push(Box::new(self.parse_italic(bounds)))
            } else if self.accept(&String::from("__")).is_some() || self.accept(&String::from("**")).is_some() {
                retval.push(Box::new(self.parse_bold(bounds)))
            } else if self.accept(&String::from("`")).is_some() {
                retval.push(Box::new(self.parse_code()))
            } else if self.accept(&String::from("[")).is_some() {
                retval.push(Box::new(self.parse_link()))
            } else {
                retval.push(Box::new(Text{text:self.pop().data}))
            }
        }
        retval
    }

    fn parse_italic(&mut self, bounds:&HashSet<String>)->Italic {
        let mut new_bounds = bounds.clone();
        new_bounds.insert(String::from("*"));
        new_bounds.insert(String::from("_"));
        let children = self.parse_formatted_text(&new_bounds);
        if self.accept(&String::from("*")).is_none() {
            self.expect(String::from("_"));
        }
        Italic{children}
    }

    fn parse_bold(&mut self, bounds:&HashSet<String>)->Bold {
        let mut new_bounds = bounds.clone();
        new_bounds.insert(String::from("**"));
        new_bounds.insert(String::from("__"));
        let children = self.parse_formatted_text(&new_bounds);
        if self.accept(&String::from("**")).is_none() {
            self.expect(String::from("__"));
        }
        Bold{children}
    }

    fn parse_code(&mut self)->Code {
        let mut text = String::from("");
        while self.accept(&String::from("`")).is_none() {
            text += &self.pop().data;
        }
        Code{text}
    }

    fn parse_link(&mut self)->Link{
        let text = self.pop().data;
        self.expect(String::from("]"));
        self.expect(String::from("("));
        let url = self.pop().data;
        self.expect(String::from(")"));
        Link{text, url}
    }
}

fn is_special_char(c:char)->bool {
    return c == '_' ||
		c == '*' ||
		c == '`' ||
		c == '#' ||
		c == '[' ||
		c == ']' ||
		c == '(' ||
		c == ')' ||
		c == '!';
}